fleetspeak/src/server/internal/cache/cache.go

```diff
@@ -0,0 +1,114 @@
+package cache
+
+import (
+	"sync"
+	"time"
+
+	"github.com/google/fleetspeak/fleetspeak/src/common"
+	"github.com/google/fleetspeak/fleetspeak/src/server/db"
+)
+
+// Controls cache timing behavior, variable only to support unit tests.
+var (
+	maxAge         = 30 * time.Second
+	expireInterval = 5 * time.Minute
+)
+
+// Client is a cache of recently connected clients.
+type Client struct {
+	m    map[common.ClientID]*clientEntry
+	l    sync.RWMutex
+	stop chan struct{}
+}
+
+type clientEntry struct {
+	u time.Time
+	d *db.ClientData
+}
+
+// NewClient returns a new cache of client data.
+func NewClient() *Client {
+	ret := &Client{
+		m:    make(map[common.ClientID]*clientEntry),
+		stop: make(chan struct{}),
+	}
+	go ret.expireLoop()
+	return ret
+}
+
+// Get returns the cached client data, if there is sufficiently fresh data in
+// the cache, otherwise nil.
+func (c *Client) Get(id common.ClientID) *db.ClientData {
+	c.l.RLock()
+	defer c.l.RUnlock()
+
+	e := c.m[id]
+	if e == nil || db.Now().Sub(e.u) > maxAge {
+		return nil
+	}
+	return e.d
+}
+
+// Update updates or sets the cached data for a particular client. If data is
+// nil, it clears the data for the client.
+func (c *Client) Update(id common.ClientID, data *db.ClientData) {
```

Comment:
This is only called in one place, with data=nil. Iiuc, that means the cache is always empty.

Reply:
Good catch. Of course commsContext.GetClientInfo should update the cache on load. Fixed.